Remove sign in when authenticating already authorized apps.

Acidity opened the pull request:
I'm submitting this to start a discussion on whether it'd be worthwhile and to raise any security concerns that it might cause. In particular, given that there have been reports of session issues with the forums, I want to make 100% certain that there would be no such issues with Core. Additionally, it might be worth making sessions expire when someone tries to use that session id from a different IP address, though this will cause shortened session times for those with dynamic IP addresses, or that switch the location of their computer often.

Acidity committed:
Remove sign in when authenticating already authorized apps.

When a user has an existing Core session, if they attempt to
authenticate an app they've already authorized, they will no longer be
prompted for their username and password. This reduces the amount that
users need to type their username and password while ensuring that they
still must do so before authorizing an application.

Signed-off-by: Tyler O'Meara <Tyler@TylerOMeara.com>

jboning commented:
I am in favor of doing this. Definitely worth thinking carefully about before we make the change though

Acidity committed:
Invalidate sessions when accessed from an IP other than the original.

Advantages of this change include increased security, as even if the
session token is leaked a hostile principal would still need to know
the IP address for it, and successfully impersonate that IP address.
Because the IP address storage is stored in the session rather than
based on the session's user's stored IP address, user's still retain
the ability to have multiple sessions concurrently. Disadvantages
include shortened, unpredictable session times for users without
static IP addresses, as well poor session support for mobile devices
that switch between Internet providers often.

Signed-off-by: Tyler O'Meara <Tyler@TylerOMeara.com>

Acidity committed:
Add domain-wide cookie for storing user authorized apps.

Core now stores the id of every app authorized by the user in a cookie,
allowing applications to determine if they should show the user an
index page with the option to authenticate, or if the application
should just make the authorize call and then redirect the user to Core
to finish the authorization. This allows for seamless SSO features
without having to share sessions between applications, allowing for
untrusted applications to partake in the SSO environment.

Signed-off-by: Tyler O'Meara <Tyler@TylerOMeara.com>

Acidity committed:
Removes implicit extension of authorizations when user does not provide credentials.

Signed-off-by: Tyler O'Meara <Tyler@TylerOMeara.com>

jboning commented on brave/core/controller.py:
```diff
@@ -35,10 +35,17 @@ def test(self):
+
+                raise HTTPFound(location='/account/authenticate?redirect=%2Fauthorize%2F{0}'.format(ar.id))
+            except AuthenticationRequest.DoesNotExist:
+                raise HTTPFound(location='/account/authenticate?redirect=%2Fauthorize%2F{0}'.format(ar.id))
```
`ar` won't be an object at this point

jboning commented on brave/core/controller.py:
```diff
@@ -154,6 +166,20 @@ class RootController(StartupMixIn, Controller):
 
     def __call__(self, req):
+
+        if session and ('host' not in session or session['host'] != request.remote_addr):
```
Hrm, remote addr is not super reliable, people with laptops move around and home dhcp leases change. Would it be okay to let people change IPs?

jboning commented on brave/core/controller.py:
```diff
@@ -154,6 +166,20 @@ class RootController(StartupMixIn, Controller):
+            authorized_apps = ', '.join([str(g.application.id) for g in user.grants])
+            if request.cookies.get('authorized_apps') != authorized_apps:
+                sso_domain = config['core.sso.domain'] if config['core.sso.domain'] != 'localhost' else None
+                response.set_cookie('authorized_apps', authorized_apps, domain=sso_domain, overwrite=True)
```
What's the goal of this, again? It was so applications can message to the user that they'll need to log in?

Acidity replied:
Yeah, right now if we just stick to the original commit, the user still needs to be prompted for and press the authorize/login button on the app. This cookie allows for Core to signal to the apps that the user has this app authorized, so they can automatically conduct the login flow so that applications automatically log you in if you have a valid Core account.

jboning commented:
Merged two commits from this. I'm downright rejecting the session IP binding, a shared cookie is probably okay but I want to think about it as a separate PR :)
